File: graham.py
"""Implémentation du parcours de Graham en python 3.
Contient les définitions nécessaire au fonctionnement de l'algorithme, la fonction enveloppe_convexe et un affichage graphique
avec Tkinter pour tester de manière interactive ladite fonction."""

class Vecteur:
    """Représentation de vecteurs en 2 dimensions. Permet de comparer entre deux vecteurs, lequel est le plus à droite."""
    def __init__(self, point_a, point_b):
        """Nouveau vecteur en fonction de 2 couples de coordonnées de points."""
        self.x = point_b[0] - point_a[0]
        self.y = point_b[1] - point_a[1]

    def det(self, other):
        """Déterminant de deux vecteurs."""
        return self.x * other.y - self.y * other.x

    def normcarr(self):
        """Le carré de la norme de self."""
        return self.x ** 2 + self.y ** 2

    def __lt__(self, other):
        """self < other : renvoie True si le vecteur self est « plus à
        gauche » que le vecteur other ou que si les deux sont colinéaires,
        si self a une norme plus petite que other."""
        det = self.det(other)
        if det < 0:
            return True
        if det > 0:
            return False
        return self.normcarr() < other.normcarr()
    
def enveloppe_convexe(points):
    """Renvoie la liste des sommets de l'enveloppe convexe d'une liste de
    points du plan, en utilisant le parcours de Graham."""
    if len(points) <= 3:
        return points.copy()

    #on recherche le pivot, c'est-à-dire le point de plus petite abscisse et de
    #plus petite ordonnée à défaut.
    pivot = points[0]
    for p in points:
        if p[0] < pivot[0] or (p[0] == pivot[0] and p[1] < pivot[1]):
            pivot = p

    points.sort(key=lambda p: Vecteur(pivot, p))
    #Tri des points en fonction de leur angle.

    pile = [pivot, points[1]]
    for i in range(2, len(points)):
        p = points[i]
        while len(pile) >= 2 and \
              Vecteur(pile[-2], p).det(Vecteur(pile[-2], pile[-1])) <= 0:
                  #Si les points pile[-2], pile[-1], p forment un tour à droite :
            pile.pop()
        pile.append(p)
 
    return pile

#Test avec Tkinter :
if __name__ == "__main__":
    import tkinter as tk
    from itertools import chain
    fenetre = tk.Tk()
    
    c = tk.Canvas(fenetre, height=800, width=800, bg="white")
    c.pack()

    points = []
    enveloppe = []
    num = None
    
    def ajouter_point(event):
        global num, enveloppe
        points.append((event.x, event.y))
        c.create_oval(event.x-3, event.y-3, event.x+3, event.y+3, fill="black")
        enveloppe = enveloppe_convexe(points)
        if num is None:
            num = c.create_polygon(*chain(*enveloppe), fill="white",
                                   outline="black", width=1.5)
        else:
            c.coords(num, *chain(*enveloppe))

    c.bind("<Button-1>", ajouter_point)

    def reset():
        global num, points, enveloppe
        c.delete(tk.ALL)
        points = []
        enveloppe = []
        num = None
        
    b = tk.Button(fenetre, text="reset", command=reset)
    b.pack()

    fenetre.mainloop()

File: graham.ml


(*mini : ('a -> 'a -> bool) -> 'a list -> 'a
  mini cmp liste : le mini mum de liste, avec cmp
  une fonction de comparaison.*)
let mini cmp liste =
  (* min' : 'a -> 'a list -> 'a
  Le mini mum entre r et le mini mun de liste.*)
  let rec min' r liste = match liste with
    |[] -> r
    |t :: q -> if (cmp t r < 0) then min' t q else min' r q
  in match liste with
    |[] -> failwith "Liste vide."
    |t :: q -> min' t q
;;

let compare_hauteur (x, y) (x', y') =
  if y > y' then 1
  else if y < y' then -1
  else compare x x'
;;

(*determinant : int * int -> int * int -> int * int -> int
  determinant O A B : le déterminant de vecteurs OA et OB.
  positif si OA est à droite par rapport à OB ;
  négatif si OA est à gauche par rapport à OB.*)
let determinant (xa, ya) (xb, yb) (xc, yc) =
  (xb - xa) * (yc - ya) - (yb - ya) * (xc - xa)
;;

let norme_carre (x, y) (x', y') =
  let dx = x - x' in
  let dy = y - y' in
  dx * dx + dy * dy
;;

(* compare_angle : int * int -> int * int -> int * int -> int
  compare_angle O A B : compare si OA est plus à droite que OB, ou
  si les deux sont colinéaires et que ||OA|| > ||OB||*)
let compare_angle o a b =
  let det = determinant o a b in
  if det < 0 then 1
  else if det > 0 then -1
  else compare (norme_carre o a) (norme_carre o b)
;;

(*graham int * int list -> int * int list
  graham l : la liste des sommets de l'enveloppe convexe des
points de l, en partant du point le plus bas.*)
let graham liste =
  if liste = [] then [] else
  let p = mini compare_hauteur liste in
  let sliste = List.sort (compare_angle1 p) liste in
  (*add_point : int * int list -> int * int -> int * int list
    add_point evlp p -> si evlp est une enveloppe convexe de points
    p1, ..., pn et que p est à droite de tous ces points, alors renvoie
    l'enveloppe convexe des points p1, ..., pn, p.*)
  let rec add_point evlp p = match evlp with
    |b :: a :: q ->
      if determinant a p b >= 0 then add_point (a :: q) p else p :: evlp
      (*S'il y a un tour à droite, alors on retire le dernier point de
      l'enveloppe et on continue la recherche. Sinon, on rajoute p à
      l'enveloppe.*)
    |_ -> p :: evlp
  in
  List.fold_left add_point [] sliste
  (*On part d'une enveloppe vide, et on ajoute les points de sliste un par un.*)
;;

(*Tests :*)
let print_liste liste =
  let rec string_of_intlist = function
    |[] -> "Nil"
    |[a, b] -> "(" ^ (string_of_int a) ^ "," ^ (string_of_int b) ^ ")"
    |(a, b) :: q ->  "(" ^ (string_of_int a) ^ "," ^ (string_of_int b) ^ ")"
^ " " ^ (string_of_intlist q)
  in
  print_endline (string_of_intlist liste)
;;

print_liste (graham []);;
print_liste (graham [(0, 0); (0, 1); (1, 0); (1, 1)]);;
print_liste (graham [(0, 0); (9, 1); (7, 3); (5, 4); (2, 6);
                                (0, 12)]);;
